fenix/app/src/main/java/org/mozilla/fenix/extension/WebExtensionPromptFeature.kt

```diff
@@ -0,0 +1,165 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.fenix.extension
+
+import android.content.Context
+import android.view.Gravity
+import android.view.View
+import androidx.annotation.VisibleForTesting
+import androidx.fragment.app.FragmentManager
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.cancel
+import kotlinx.coroutines.flow.mapNotNull
+import mozilla.components.browser.state.action.WebExtensionAction
+import mozilla.components.browser.state.state.extension.WebExtensionPromptRequest
+import mozilla.components.browser.state.store.BrowserStore
+import mozilla.components.feature.addons.Addon
+import mozilla.components.feature.addons.ui.PermissionsDialogFragment
+import mozilla.components.lib.state.ext.flowScoped
+import mozilla.components.support.base.feature.LifecycleAwareFeature
+import mozilla.components.support.ktx.kotlinx.coroutines.flow.ifChanged
+import org.mozilla.fenix.R
+import org.mozilla.fenix.addons.showSnackBar
+import org.mozilla.fenix.components.FenixSnackbar
+import org.mozilla.fenix.theme.ThemeManager
+
+/**
+ * Feature implementation for handling [WebExtensionPromptRequest] and showing the respective UI.
+ */
+class WebExtensionPromptFeature(
+    private val store: BrowserStore,
+    private val provideAddons: suspend () -> List<Addon>,
+    private val context: Context,
+    private val view: View,
+    private val fragmentManager: FragmentManager,
+) : LifecycleAwareFeature {
+
+    /**
+     * Whether or not an add-on installation is in progress.
+     */
+    private var isInstallationInProgress = false
+    private var scope: CoroutineScope? = null
+
+    /**
+     * Starts observing the selected session to listen for window requests
+     * and opens / closes tabs as needed.
+     */
+    override fun start() {
+        scope = store.flowScoped { flow ->
+            flow.mapNotNull { state ->
+                state.webExtensionPromptRequest
+            }.ifChanged().collect { promptRequest ->
+                if (promptRequest is WebExtensionPromptRequest.Permissions && !hasExistingPermissionDialogFragment()) {
+                    val addon = provideAddons().find { addon ->
+                        addon.id == promptRequest.extension.id
+                    }
+
+                    if (addon == null) {
```

Comment:
In the future, this might not be because the add-on is already installed but because the install process was initiated for an add-on that isn't "known" to Fenix. I am not sure we'll run into that corner case here, though, so I'd suggest we document the current assumption:

```suggestion
                    // If the add-on is not found, it is already installed because the install process can only
                    // be triggered for add-ons "known" by Fenix (the add-on is either part of the official list
                    // of supported extensions OR part of the user custom AMO collection).
                    if (addon == null) {
```